### PR TITLE
Fixed #10929 -- Support a default value for Sum

### PR DESCRIPTION
**Pull request by bartonip**
Added default=None keyword argument to Sum

Added tests for each supported output_field type

Update docs/topics/db/aggregation.txt

https://code.djangoproject.com/ticket/10929

### CODE REVIEW

**Review by felixxm — COMMENTED** [review]
@bartonip Thanks for this patch :+1: What do you think about [Simon's proposition](https://code.djangoproject.com/ticket/10929#comment:20) to add `coalesce` instead of default?

Also, release notes and `versionchanged` annotations are missing.

**Comment by bartonip** [comment]
@felixxm I think Simon's idea is great, however I think the more likely search terms to reach/search for when trying to solve this problem would be something to do with "default value" or "base value". 

I am definitely happy to include support for expressions in this patch, but I think the kwarg name should still be default.

**Comment by bartonip** [comment]
Adding onto my previous comment, the pattern for a default value for fields is `default` so I think continuing this pattern would make it easier for those who are not familiar with the framework to find the documentation for this argument. Especially if you do not have too much SQL experience, you wouldn't know what `coalesce` does.

**Comment by bartonip** [comment]
@felixxm The issues have been fixed.

**Review by felixxm — COMMENTED** [review]
@bartonip Thanks :+1: Please check tests that don't work on PostgreSQL and Oracle.

File: django/db/models/aggregates.py [hunk 1]
```diff
@@ -145,6 +145,19 @@ class Sum(FixDurationInputMixin, Aggregate):
...
+    def __init__(self, *expressions, default=None, **extra):
+        self.default_value = default
+        super().__init__(*expressions, **extra)
+
+    def convert_value(self, value, expression, connection):
+        value = self.default_value if value is None else value
+
+        import datetime
+        if isinstance(value, datetime.timedelta):
+            value = value.microseconds
```
Comment (pope1ni) [inline django/db/models/aggregates.py:157]: I don't understand why this is here? Also why not `value.total_seconds()`? Or am I missing something?

```python
>>> from datetime import timedelta
>>> delta = timedelta(microseconds=90000000123)
>>> delta
datetime.timedelta(days=1, seconds=3600, microseconds=123)
>>> delta.microseconds
123
>>> delta.total_seconds()
90000.000123
```

**Review by pope1ni — CHANGES_REQUESTED** [review]
I don't think this is the correct approach. If the `default` argument is not `None` we should wrap the function in `as_sql()` with `Coalesce()`. This is important for allowing expressions such as the following:

- `Sum('rating', default=F('other'))`
- `Sum('rating', default=0) + Value(5)`

Basically we want the following:

- `Sum('rating', default=None)` → `SUM(rating)`
- `Sum('rating', default=0) + Value(5)` → `COALESCE(SUM(rating), 0) + 5`

This is essentially what @charettes has been proposing on the ticket. I do, however, agree that `default` as an argument name is probably more discoverable than `coalesce` and we can document that it wraps the aggregate function with `COALESCE` to replace `None` with the specified default.

We should also do this in the `Aggregate` base class rather than for `Sum` only, as this issue also applies to `Avg`, etc. This was also mentioned in the ticket.

At the end of the day this is merely syntactic sugar for wrapping the expression with `COALESCE` which is the way it'd be handled if you were writing raw SQL, but it is worth having as if is a very common use case and something that people seem to have trouble understanding.

**Review by recepsirin — APPROVED** [review]

**Comment by pope1ni** [comment]
Superseded by #14026.

**Pull request closed without merging**